File: ASP.NET MVC Editores Date y DateTime con helper para gestión de scripts/[VB]-ASP.NET MVC Editores Date y DateTime con helper para gestión de scripts/VB/MVCDateTimePicker/Infrastructure/ScriptHelper.vb
Imports System.Web.Mvc
Imports System.Runtime.CompilerServices

Public Module ScriptHelper

    Private Const ScriptItemsKey As String = "ScriptBlocks"

    Private Enum ScriptType
        Code
        File
    End Enum

    Private Class ScriptBlock
        Property ScriptType As ScriptType
        Property Script As String
    End Class

    Private Function ScriptCollection(html As HtmlHelper) As Dictionary(Of String, ScriptBlock)
        Dim _httpContext As HttpContextBase = html.ViewContext.HttpContext
        If Not _httpContext.Items.Contains(ScriptItemsKey) Then
            _httpContext.Items(ScriptItemsKey) = New Dictionary(Of String, ScriptBlock)
        End If
        Return CType(_httpContext.Items(ScriptItemsKey), Dictionary(Of String, ScriptBlock))
    End Function

    <Extension> _
    Public Function AddScript(html As HtmlHelper, key As String, scriptCode As String) As MvcHtmlString
        Dim scripts As Dictionary(Of String, ScriptBlock) = ScriptCollection(html)
        If Not scripts.ContainsKey(key) Then
            scripts.Add(key _
                        , New ScriptBlock() With {.ScriptType = ScriptType.Code, .Script = scriptCode})
        End If
        Return MvcHtmlString.Empty
    End Function

    <Extension> _
    Public Function AddScriptFile(html As HtmlHelper, key As String, scriptFile As String) As MvcHtmlString
        Dim scripts As Dictionary(Of String, ScriptBlock) = ScriptCollection(html)
        If Not scripts.ContainsKey(key) Then
            scripts.Add(key _
                        , New ScriptBlock() With {.ScriptType = ScriptType.File, .Script = scriptFile})
        End If
        Return MvcHtmlString.Empty
    End Function

    <Extension> _
    Public Function RenderScripts(html As HtmlHelper) As MvcHtmlString
        Dim scripts As Dictionary(Of String, ScriptBlock) = ScriptCollection(html)
        Dim result As String = String.Empty
        For Each item As ScriptBlock In scripts.Values
            Dim scriptTag As TagBuilder = New TagBuilder("script")
            scriptTag.MergeAttribute("type", "text/javascript")
            If item.ScriptType = ScriptType.File Then
                scriptTag.MergeAttribute("src", UrlHelper.GenerateContentUrl(item.Script, html.ViewContext.HttpContext))
            Else
                scriptTag.InnerHtml = Environment.NewLine & item.Script & Environment.NewLine
            End If
            result = result & scriptTag.ToString(TagRenderMode.Normal) & Environment.NewLine
        Next
        html.ViewContext.HttpContext.Items.Remove(ScriptItemsKey)
        Return MvcHtmlString.Create(result)
    End Function

End Module

File: ASP.NET MVC Editores Date y DateTime con helper para gestión de scripts/[C#]-ASP.NET MVC Editores Date y DateTime con helper para gestión de scripts/C#/MVCDateTimePicker/Infrastructure/ScriptHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace MVCDateTimePicker.Infrastructure
{
    public static class ScriptHelper
    {

        private const string scriptItemsKey = "ScriptBlocks";

        private enum ScriptType { Code, File }

        private class ScriptBlock
        {
            public ScriptType ScriptType { get; set; }
            public string Script { get; set; }
        }

        private static Dictionary<string, ScriptBlock> ScriptCollection(HtmlHelper html)
        {
            HttpContextBase _httpContext = html.ViewContext.HttpContext;
            if (!_httpContext.Items.Contains(scriptItemsKey))
                _httpContext.Items[scriptItemsKey] = new Dictionary<string, ScriptBlock>();
            return (Dictionary<string, ScriptBlock>)_httpContext.Items[scriptItemsKey];
        }

        public static MvcHtmlString AddScript(this HtmlHelper html, string key, string scriptCode)
        {
            Dictionary<string, ScriptBlock> scripts = ScriptCollection(html);
            if (!scripts.ContainsKey(key))
                scripts.Add(key
                    , new ScriptBlock { ScriptType = ScriptType.Code, Script = scriptCode });
            return MvcHtmlString.Empty;
        }

        public static MvcHtmlString AddScriptFile(this HtmlHelper html, string key, string scriptFile)
        {
            Dictionary<string, ScriptBlock> scripts = ScriptCollection(html);
            if (!scripts.ContainsKey(key))
                scripts.Add(key
                , new ScriptBlock { ScriptType = ScriptType.File, Script = scriptFile });
            return MvcHtmlString.Empty;
        }

        public static MvcHtmlString RenderScripts(this HtmlHelper html)
        {
            Dictionary<string, ScriptBlock> scripts = ScriptCollection(html);
            string result = string.Empty;
            foreach (ScriptBlock item in scripts.Values)
            {
                TagBuilder scriptTag = new TagBuilder("script");
                scriptTag.MergeAttribute("type", "text/javascript");
                if (item.ScriptType == ScriptType.File)
                    scriptTag.MergeAttribute("src", UrlHelper.GenerateContentUrl(item.Script, html.ViewContext.HttpContext));
                else
                    scriptTag.InnerHtml = Environment.NewLine + item.Script + Environment.NewLine;
                result += scriptTag.ToString(TagRenderMode.Normal) + Environment.NewLine;
            }
            html.ViewContext.HttpContext.Items.Remove(scriptItemsKey);
            return MvcHtmlString.Create(result);
        }
    }
}